<a href="https://colab.research.google.com/github/alouvre/Machine-Learning-Projects/blob/main/CollaborativeFiltering.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **AKSES KE GOOGLE DRIVE**
---

In [4]:
# from google.colab import drive
# drive.mount('/content/drive')

# **AKSES DATA DARI REPO GITHUB**
---

In [5]:
!git clone https://github.com/codeheroku/Introduction-to-Machine-Learning.git

Cloning into 'Introduction-to-Machine-Learning'...
remote: Enumerating objects: 537, done.
remote: Counting objects: 100% (8/8), done.
remote: Compressing objects: 100% (6/6), done.
remote: Total 537 (delta 3), reused 6 (delta 2), pack-reused 529
Receiving objects: 100% (537/537), 24.07 MiB | 7.64 MiB/s, done.
Resolving deltas: 100% (172/172), done.
Updating files: 100% (356/356), done.


In [6]:
cd Introduction-to-Machine-Learning/Collaborative Filtering/dataset

/content/drive/MyDrive/Recommendation_System/Introduction-to-Machine-Learning/Collaborative Filtering/dataset


# **COLLABORATIVE FILTERING**
---

Collaborative filtering adalah salah satu metode yang digunakan dalam sistem rekomendasi untuk menghasilkan rekomendasi berdasarkan pola perilaku pengguna atau item yang mirip. Metode ini memang seringkali mencocokkan dua variabel, yaitu pengguna (user) dan item (produk atau konten), tetapi tidak hanya berdasarkan nilai korelasinya.

### Ada dua jenis utama dari collaborative filtering:



#### 1. User-Based Collaborative Filtering
Metode ini mencari pengguna lain yang memiliki preferensi serupa dengan pengguna yang sedang mencari rekomendasi. Ini dilakukan dengan membandingkan riwayat peringkat atau interaksi pengguna dengan item. Pengguna yang memiliki pola peringkat serupa dengan pengguna yang sedang mencari rekomendasi dianggap memiliki preferensi yang mirip. Rekomendasi kemudian dibuat berdasarkan apa yang disukai oleh pengguna-pengguna tersebut.

#### 2. Item-Based Collaborative Filtering
Metode ini mencocokkan item-item yang mirip berdasarkan sejarah peringkat pengguna. Jika sejumlah besar pengguna memberikan peringkat yang serupa pada dua item, maka item-item tersebut dianggap mirip. Rekomendasi kemudian dibuat berdasarkan item yang mirip dengan item yang sedang dicari oleh pengguna.

Dalam kedua jenis collaborative filtering di atas, perhitungan korelasi antara pengguna atau item memang dapat digunakan sebagai salah satu metode untuk menentukan kesamaan. Salah satu metrik yang umum digunakan adalah korelasi Pearson, yang mengukur hubungan linear antara dua variabel. Namun, ada juga metrik lain yang dapat digunakan, seperti jarak kosinus, indeks Jaccard, atau metode lainnya.

# **EXAMPLE**
---

Collaborative filtering adalah teknik yang digunakan dalam sistem rekomendasi untuk memberikan rekomendasi berdasarkan perilaku atau preferensi pengguna. Saya akan memberikan contoh dua jenis utama collaborative filtering, yaitu User-Based Collaborative Filtering dan Item-Based Collaborative Filtering, menggunakan dataset sederhana sebagai ilustrasi.

## Contoh 1

Misalkan Anda memiliki dataset peringkat film oleh beberapa pengguna. Dataset ini mungkin terlihat seperti ini:

In [7]:
import pandas as pd
import numpy as np

In [8]:
# membuat data
data_ratings = {
    'User_ID' : ['User_1', 'User_2', 'User_3', 'User_4'],
    'Movie_1' : [5, np.nan, 3, 4],
    'Movie_2' : [4, 5, np.nan, 2],
    'Movie_3' : [np.nan, 4, np.nan, 3]
}
data_ratings

{'User_ID': ['User_1', 'User_2', 'User_3', 'User_4'],
 'Movie_1': [5, nan, 3, 4],
 'Movie_2': [4, 5, nan, 2],
 'Movie_3': [nan, 4, nan, 3]}

In [9]:
# membuat dataframe berdasarkan data
df = pd.DataFrame(data_ratings)
# menampilkan dataframe df
df

User_ID  Movie_1  Movie_2  Movie_3
0  User_1      5.0      4.0      NaN
1  User_2      NaN      5.0      4.0
2  User_3      3.0      NaN      NaN
3  User_4      4.0      2.0      3.0

Dalam User-Based Collaborative Filtering, kita akan mencari pengguna lain yang memiliki preferensi serupa dengan pengguna yang sedang mencari rekomendasi. Misalnya, jika User_1 menyukai Movie_1 dan Movie_2, kita dapat mencari pengguna lain yang juga memberikan peringkat tinggi untuk Movie_1 dan Movie_2. Jika User_3 memberikan peringkat tinggi untuk Movie_1, kita dapat mengasumsikan bahwa User_1 dan User_3 memiliki preferensi serupa.

# **CASE - Movie Recomendation**
---

In [10]:
import pandas as pd
from scipy import sparse

In [11]:
# inisialisasi data rating
ratings = pd.read_csv('ratings.csv')
# inisialisasi data movies
movies = pd.read_csv('movies.csv')
# mengetahui dimensi data ratings
print(ratings.shape)
# menampilkan 5 baris pertama dari data rating
ratings.head()

(100836, 4)


userId  movieId  rating  timestamp
0       1        1     4.0  964982703
1       1        3     4.0  964981247
2       1        6     4.0  964982224
3       1       47     5.0  964983815
4       1       50     5.0  964982931

In [12]:
# mengetahui dimensi dari data movies
print(movies.shape)
# menampilkan 5 baris pertama
movies.head()

(9742, 3)


movieId                               title  \
0        1                    Toy Story (1995)   
1        2                      Jumanji (1995)   
2        3             Grumpier Old Men (1995)   
3        4            Waiting to Exhale (1995)   
4        5  Father of the Bride Part II (1995)   

                                        genres  
0  Adventure|Animation|Children|Comedy|Fantasy  
1                   Adventure|Children|Fantasy  
2                               Comedy|Romance  
3                         Comedy|Drama|Romance  
4                                       Comedy

In [13]:
# drop kolom genres dan timestamp
# kolom yang akan digunakan kolom movieId, title, userId, dan rating
ratings = pd.merge(movies, ratings).drop(['genres','timestamp'],axis=1)
# menampilkan 5 baris pertama
ratings.head()

movieId             title  userId  rating
0        1  Toy Story (1995)       1     4.0
1        1  Toy Story (1995)       5     4.0
2        1  Toy Story (1995)       7     4.5
3        1  Toy Story (1995)      15     2.5
4        1  Toy Story (1995)      17     4.5

In [14]:
# membuat pivot table
# jadi ini data pemberian rating pada film oleh user
# kolomnya nama-nama film
# valuenya rating tiap film
userRatings = ratings.pivot_table(index=['userId'],
                                  columns=['title'],
                                  values='rating')
# menampilkan dimensi sebelum perubahan
print("Before: ",userRatings.shape)
# menampilkan 5 baris pertama
userRatings.head()

Before:  (610, 9719)


title   '71 (2014)  'Hellboy': The Seeds of Creation (2004)  \
userId                                                        
1              NaN                                      NaN   
2              NaN                                      NaN   
3              NaN                                      NaN   
4              NaN                                      NaN   
5              NaN                                      NaN   

title   'Round Midnight (1986)  'Salem's Lot (2004)  \
userId                                                
1                          NaN                  NaN   
2                          NaN                  NaN   
3                          NaN                  NaN   
4                          NaN                  NaN   
5                          NaN                  NaN   

title   'Til There Was You (1997)  'Tis the Season for Love (2015)  \
userId                                                               
1                             NaN                              NaN   
2                             NaN                              NaN   
3                             NaN                              NaN   
4                             NaN                              NaN   
5                             NaN                              NaN   

title   'burbs, The (1989)  'night Mother (1986)  (500) Days of Summer (2009)  \
userId                                                                          
1                      NaN                   NaN                          NaN   
2                      NaN                   NaN                          NaN   
3                      NaN                   NaN                          NaN   
4                      NaN                   NaN                          NaN   
5                      NaN                   NaN                          NaN   

title   *batteries not included (1987)  ...  Zulu (2013)  [REC] (2007)  \
userId                                  ...                              
1                                  NaN  ...          NaN           NaN   
2                                  NaN  ...          NaN           NaN   
3                                  NaN  ...          NaN           NaN   
4                                  NaN  ...          NaN           NaN   
5                                  NaN  ...          NaN           NaN   

title   [REC]² (2009)  [REC]³ 3 Génesis (2012)  \
userId                                           
1                 NaN                      NaN   
2                 NaN                      NaN   
3                 NaN                      NaN   
4                 NaN                      NaN   
5                 NaN                      NaN   

title   anohana: The Flower We Saw That Day - The Movie (2013)  \
userId                                                           
1                                                     NaN        
2                                                     NaN        
3                                                     NaN        
4                                                     NaN        
5                                                     NaN        

title   eXistenZ (1999)  xXx (2002)  xXx: State of the Union (2005)  \
userId                                                                
1                   NaN         NaN                             NaN   
2                   NaN         NaN                             NaN   
3                   NaN         NaN                             NaN   
4                   NaN         NaN                             NaN   
5                   NaN         NaN                             NaN   

title   ¡Three Amigos! (1986)  À nous la liberté (Freedom for Us) (1931)  
userId                                                                    
1                         4.0                                        NaN  
2                         NaN                            

In [15]:
# drop nilai yang bukan NaN yang jumlahnya kurang dari 10
# replace NaN menjadi 0
userRatings = userRatings.dropna(thresh=10, axis=1).fillna(0,axis=1)
# menampilkan dimensi setelah perubahan
print("After: ",userRatings.shape)
# menampilkan 5 baris pertama
userRatings.head()

After:  (610, 2269)


title   'burbs, The (1989)  (500) Days of Summer (2009)  \
userId                                                    
1                      0.0                          0.0   
2                      0.0                          0.0   
3                      0.0                          0.0   
4                      0.0                          0.0   
5                      0.0                          0.0   

title   10 Cloverfield Lane (2016)  10 Things I Hate About You (1999)  \
userId                                                                  
1                              0.0                                0.0   
2                              0.0                                0.0   
3                              0.0                                0.0   
4                              0.0                                0.0   
5                              0.0                                0.0   

title   10,000 BC (2008)  101 Dalmatians (1996)  \
userId                                            
1                    0.0                    0.0   
2                    0.0                    0.0   
3                    0.0                    0.0   
4                    0.0                    0.0   
5                    0.0                    0.0   

title   101 Dalmatians (One Hundred and One Dalmatians) (1961)  \
userId                                                           
1                                                     0.0        
2                                                     0.0        
3                                                     0.0        
4                                                     0.0        
5                                                     0.0        

title   12 Angry Men (1957)  12 Years a Slave (2013)  127 Hours (2010)  ...  \
userId                                                                  ...   
1                       0.0                      0.0               0.0  ...   
2                       0.0                      0.0               0.0  ...   
3                       0.0                      0.0               0.0  ...   
4                       5.0                      0.0               0.0  ...   
5                       0.0                      0.0               0.0  ...   

title   Zack and Miri Make a Porno (2008)  Zero Dark Thirty (2012)  \
userId                                                               
1                                     0.0                      0.0   
2                                     0.0                      0.0   
3                                     0.0                      0.0   
4                                     0.0                      0.0   
5                                     0.0                      0.0   

title   Zero Effect (1998)  Zodiac (2007)  Zombieland (2009)  \
userId                                                         
1                      0.0            0.0                0.0   
2                      0.0            0.0                3.0   
3                      0.0            0.0                0.0   
4                      0.0            0.0                0.0   
5                      0.0            0.0                0.0   

title   Zoolander (2001)  Zootopia (2016)  eXistenZ (1999)  xXx (2002)  \
userId                                                                   
1                    0.0              0.0              0.0         0.0   
2                    0.0              0.0              0.0         0.0   
3                    0.0              0.0              0.0         0.0   
4                    0.0              0.0              0.0         0.0   
5                    0.0              0.0              0.0         0.0   

title   ¡Three Amigos! (1986)  
userId                         
1                         4.0  
2                         0.0  
3                         0.0  
4                         0.0  
5                         0.0  

[5 rows x 2269 

In [16]:
# mengetahui korelasi antara semua pasangan kolom dalam dataframe
# metode pearson ini untuk mengukur hubungan linear antara 2 variabel
corrMatrix = userRatings.corr(method='pearson')
# menampilkan 100 baris pertama
corrMatrix.head(100)

title                              'burbs, The (1989)  \
title                                                   
'burbs, The (1989)                           1.000000   
(500) Days of Summer (2009)                  0.063117   
10 Cloverfield Lane (2016)                  -0.023768   
10 Things I Hate About You (1999)            0.143482   
10,000 BC (2008)                             0.011998   
...                                               ...   
Almost Famous (2000)                         0.099554   
Along Came Polly (2004)                      0.027287   
Along Came a Spider (2001)                   0.064762   
Amadeus (1984)                               0.136013   
Amazing Spider-Man, The (2012)               0.083419   

title                              (500) Days of Summer (2009)  \
title                                                            
'burbs, The (1989)                                    0.063117   
(500) Days of Summer (2009)                           1.000000   
10 Cloverfield Lane (2016)                            0.142471   
10 Things I Hate About You (1999)                     0.273989   
10,000 BC (2008)                                      0.193960   
...                                                        ...   
Almost Famous (2000)                                  0.209549   
Along Came Polly (2004)                               0.282426   
Along Came a Spider (2001)                           -0.003205   
Amadeus (1984)                                        0.084829   
Amazing Spider-Man, The (2012)                        0.224961   

title                              10 Cloverfield Lane (2016)  \
title                                                           
'burbs, The (1989)                                  -0.023768   
(500) Days of Summer (2009)                          0.142471   
10 Cloverfield Lane (2016)                           1.000000   
10 Things I Hate About You (1999)                   -0.005799   
10,000 BC (2008)                                     0.112396   
...                                                       ...   
Almost Famous (2000)                                 0.032088   
Along Came Polly (2004)                              0.113213   
Along Came a Spider (2001)                           0.016372   
Amadeus (1984)                                      -0.055707   
Amazing Spider-Man, The (2012)                       0.149903   

title                              10 Things I Hate About You (1999)  \
title                                                                  
'burbs, The (1989)                                          0.143482   
(500) Days of Summer (2009)                                 0.273989   
10 Cloverfield Lane (2016)                                 -0.005799   
10 Things I Hate About You (1999)                           1.000000   
10,000 BC (2008)                                            0.244670   
...                                                              ...   
Almost Famous (2000)                                        0.296727   
Along Came Polly (2004)                                     0.193085   
Along Came a Spider (2001)                                  0.085365   
Amadeus (1984)                                              0.105783   
Amazing Spider-Man, The (2012)                              0.103802   

title                              10,000 BC (2008)  101 Dalmatians (1996)  \
title                                                                        
'burbs, The (1989)                         0.011998               0.087931   
(500) Days of Summer (2009)                0.193960               0.148903   
10 Cloverfield Lane (2016)                 0.112396               0.006139   
10 Things I Hate About You (1999)          0.244670               0.223481   
10,000 BC (2008)                           1.000000               0.234459   
...                                             ...                   

In [17]:
# fungsi get_similar
# untuk mencari film-film yang memiliki korelasi tinggi dalam film tertentu
# berdasarkan matriks korelasi yang telah dihitung sebelumnya (dinyatakan pada corrMatrix)
def get_similar(movie_name,rating):
  # mengambil korelasi antara film tertentu dengan semua film lainnya
  similiar_ratings = corrMatrix[movie_name]*(rating-2.5)
  # mengurutkan film-film berdasarkan nilai korelasi secara menurun
  similiar_ratings = similiar_ratings.sort_values(ascending=False)
  # print(type(similiar_ratings))
  return similiar_ratings

In [18]:
# Contoh inputan dari User
# Misalkan User melakukan pencarian sebanyak 4 kali dengan kategori romantic-lover
# Daftar film-film yang dicari dan peringkat yang diberikan
romantic_lover = [("(500) Days of Summer (2009)",5),
                  ("Alice in Wonderland (2010)",3),
                  ("Aliens (1986)",1),
                  ("2001: A Space Odyssey (1968)",2)]
# membuat dataframe kosong
similiar_movies = pd.DataFrame()
# pengulangan untuk daftar romantic_lover
for movie,rating in romantic_lover:
  # memanggil fungsi get_similar
  # untuk mencari film-film yang mirip berdasarkan film dan peringkat
  similiar_movies = similiar_movies.append(get_similar(movie,rating),
                                           ignore_index=True)
# menampilkan 5 baris pertama
similiar_movies.head()

<ipython-input-18-bbdcdfe0d0df>:14: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  similiar_movies = similiar_movies.append(get_similar(movie,rating),
<ipython-input-18-bbdcdfe0d0df>:14: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  similiar_movies = similiar_movies.append(get_similar(movie,rating),
<ipython-input-18-bbdcdfe0d0df>:14: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  similiar_movies = similiar_movies.append(get_similar(movie,rating),
<ipython-input-18-bbdcdfe0d0df>:14: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  similiar_movies = similiar_movies.append(get_similar(movie,rating),


title  (500) Days of Summer (2009)  Silver Linings Playbook (2012)  \
0                         2.500000                        1.254898   
1                         0.203998                        0.191226   
2                        -0.062634                       -0.131106   
3                        -0.056808                       -0.060217   

title  Adventureland (2009)  Up in the Air (2009)  50/50 (2011)  \
0                  1.157021              1.129961      1.123890   
1                  0.130105              0.221842      0.139619   
2                 -0.091790             -0.205500     -0.112038   
3                 -0.083101             -0.093265     -0.064954   

title  Descendants, The (2011)  Crazy, Stupid, Love. (2011)  \
0                     1.082774                     1.076731   
1                     0.058567                     0.186331   
2                    -0.187131                    -0.131742   
3                    -0.059233                    -0.042563   

title  About Time (2013)  Toy Story 3 (2010)  \
0               1.050643            1.050386   
1               0.148771            0.147115   
2              -0.057615           -0.170463   
3              -0.039608           -0.063763   

title  Secret Life of Walter Mitty, The (2013)  ...  \
0                                     1.043048  ...   
1                                     0.179240  ...   
2                                    -0.160831  ...   
3                                    -0.063478  ...   

title  Remains of the Day, The (1993)  Leaving Las Vegas (1995)  \
0                           -0.127058                 -0.131261   
1                           -0.004244                 -0.022697   
2                           -0.066664                 -0.093771   
3                           -0.022207                 -0.035515   

title  Grifters, The (1990)  English Patient, The (1996)  \
0                 -0.134995                    -0.135628   
1                  0.000935                    -0.028072   
2                 -0.308606                    -0.328057   
3                 -0.144882                    -0.141102   

title  Dances with Wolves (1990)  Stargate (1994)  \
0                      -0.142231        -0.142870   
1                       0.013454        -0.010708   
2                      -0.170352        -0.329190   
3                      -0.047911        -0.078467   

title  Madness of King George, The (1994)  Disclosure (1994)  \
0                               -0.144495          -0.160575   
1                               -0.022814          -0.030627   
2                               -0.110041           0.122334   
3                               -0.020044           0.022402   

title  Postman, The (Postino, Il) (1994)  Clear and Present Danger (1994)  
0                              -0.161373                        -0.209810  
1                              -0.025479                        -0.036196  
2                              -0.113838                        -0.124716  
3                              -0.061293                        -0.023133  

[4 rows x 2269 columns]

In [19]:
# menghitung total nilai korelasi dari Dataframe similar_movies
# dilakukan pengurutan secara menurun
# dengan menampilkan 20 baris pertama
similiar_movies.sum().sort_values(ascending=False).head(20)

title
(500) Days of Summer (2009)                      2.584556
Alice in Wonderland (2010)                       1.395229
Silver Linings Playbook (2012)                   1.254800
Yes Man (2008)                                   1.116264
Adventureland (2009)                             1.112235
Marley & Me (2008)                               1.108381
About Time (2013)                                1.102192
Crazy, Stupid, Love. (2011)                      1.088757
50/50 (2011)                                     1.086517
Help, The (2011)                                 1.075963
Up in the Air (2009)                             1.053037
Holiday, The (2006)                              1.034470
Friends with Benefits (2011)                     1.030875
Notebook, The (2004)                             1.025880
Easy A (2010)                                    1.015771
Secret Life of Walter Mitty, The (2013)          0.997979
Perks of Being a Wallflower, The (2012)          0.967425
Toy Stor

In [20]:
# Contoh Inputan User
# Misalkan User melakukan pencarian 4 kali dengan kategori action-lover
# Daftar film-film yang dicari dengan peringkat yang diberikan
action_lover = [("Amazing Spider-Man, The (2012)",5),
                ("Mission: Impossible III (2006)",4),
                ("Toy Story 3 (2010)",2),
                ("2 Fast 2 Furious (Fast and the Furious 2, The) (2003)",4)]
# membuat dataframe kosong
similiar_movies = pd.DataFrame()
# pengulangan untuk daftar action_lover
for movie,rating in action_lover:
  # memanggil fungsi get_similar
  # untuk mencari film-film yang mirip berdasarkan film dan peringkat
  similiar_movies = similiar_movies.append(get_similar(movie,rating),
                                           ignore_index=True)
# menampilkan 5 baris pertama
similiar_movies.head()

<ipython-input-20-c2a3932a9dfe>:14: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  similiar_movies = similiar_movies.append(get_similar(movie,rating),
<ipython-input-20-c2a3932a9dfe>:14: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  similiar_movies = similiar_movies.append(get_similar(movie,rating),
<ipython-input-20-c2a3932a9dfe>:14: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  similiar_movies = similiar_movies.append(get_similar(movie,rating),
<ipython-input-20-c2a3932a9dfe>:14: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  similiar_movies = similiar_movies.append(get_similar(movie,rating),


title  Amazing Spider-Man, The (2012)  \
0                            2.500000   
1                            0.534632   
2                           -0.196097   
3                            0.394598   

title  Captain America: The First Avenger (2011)  Thor (2011)  \
0                                       1.370454     1.324846   
1                                       0.546741     0.534338   
2                                      -0.212966    -0.221254   
3                                       0.304620     0.364342   

title  Wreck-It Ralph (2012)  Star Trek Into Darkness (2013)  \
0                   1.302965                        1.301220   
1                   0.508677                        0.397879   
2                  -0.159085                       -0.217821   
3                   0.415351                        0.389789   

title  The Hunger Games: Mockingjay - Part 1 (2014)  \
0                                          1.298310   
1                                          0.266710   
2                                         -0.174118   
3                                          0.265608   

title  The Amazing Spider-Man 2 (2014)  X-Men: First Class (2011)  \
0                             1.297114                   1.284405   
1                             0.658657                   0.489909   
2                            -0.195072                  -0.252112   
3                             0.392979                   0.281709   

title  Tron: Legacy (2010)  Thor: The Dark World (2013)  ...  Outbreak (1995)  \
0                 1.268557                     1.250917  ...        -0.106419   
1                 0.558632                     0.590377  ...         0.142277   
2                -0.104276                    -0.183114  ...         0.030545   
3                 0.388931                     0.327142  ...         0.017307   

title  Grifters, The (1990)  Nine Months (1995)  Arsenic and Old Lace (1944)  \
0                 -0.110307           -0.114825                    -0.115754   
1                 -0.067056            0.036215                    -0.030504   
2                  0.020097            0.007077                     0.019074   
3                 -0.053135           -0.031720                    -0.055758   

title  Madness of King George, The (1994)  Disclosure (1994)  \
0                               -0.118070          -0.120780   
1                               -0.071775           0.021249   
2                                0.034244           0.039228   
3                               -0.056874          -0.025216   

title  Like Water for Chocolate (Como agua para chocolate) (1992)  \
0                                              -0.122878            
1                                              -0.074697            
2                                               0.024539            
3                                              -0.059190            

title  Postman, The (Postino, Il) (1994)  Dead Man Walking (1995)  \
0                              -0.131861                -0.134867   
1                              -0.045873                -0.000239   
2                               0.038244                 0.004534   
3                              -0.063517                -0.046472   

title  Piano, The (1993)  
0              -0.170358  
1              -0.026444  
2               0.040765  
3              -0.073868  

[4 rows x 2269 columns]

In [21]:
# menghitung total nilai korelasi dari Dataframe similar_movies
# dilakukan pengurutan secara menurun
# dengan menampilkan 20 baris pertama
similiar_movies.sum().sort_values(ascending=False).head(20)

title
Amazing Spider-Man, The (2012)                           3.233134
Mission: Impossible III (2006)                           2.874798
2 Fast 2 Furious (Fast and the Furious 2, The) (2003)    2.701477
Over the Hedge (2006)                                    2.229721
Crank (2006)                                             2.176259
Mission: Impossible - Ghost Protocol (2011)              2.159666
Hancock (2008)                                           2.156098
The Amazing Spider-Man 2 (2014)                          2.153677
Hellboy (2004)                                           2.137518
Snakes on a Plane (2006)                                 2.137396
Jumper (2008)                                            2.129716
Chronicles of Riddick, The (2004)                        2.121689
Tron: Legacy (2010)                                      2.111843
Fantastic Four (2005)                                    2.083022
X-Men: The Last Stand (2006)                             2.077530
Wrec

# **RECOMMENDATION SYSTEM**
---

User-based collaborative filtering is also called user-user collaborative filtering. It is a type recommendatioin system algorithm that uses user similarity to make product recommendations.

In this tutorial, we will talk about
- What is user-based (user-user) collaborative filtering?
- How to create a user-product matrix?
- How to process data for user-based collaborative filtering?
- How to identify similar users?
- How to narrow down the items pool?
- How to rank items for the recommendation?
- How to predict the rating score?

##**Step 0: User-Based Collaborative Filtering Recommendation Algorithm**

Firstly, let's understand how User-based collaborative filtering works.
User-based collaborative filtering makes recommendation based on user-product interactions in the past. The assumption behind algorithm is that similar users like similar products.
User-based collaborative filtering algorithm usually has the following steps:
1. Find similar users based on interactions with common items.
2. Identify the items rated high by similar users but have not been exposed to the active user of interest
3. Calculate the weight average score for each item.
4. Rank items based on the score and pick top n items to recommend

## **Step 1: Import Python Libraries**

In the first step, we will import Python libraries pandas, numpy, and scipy.stats. These three libraries are for data processing and calculations.

We also imported seaborn for visualization and cosine_similarity for calculating similarity score.

In [22]:
# Data Processing
import pandas as pd
import numpy as np
import scipy.stats

# Visualization
import seaborn as sns

# Similarity
from sklearn.metrics.pairwise import cosine_similarity

## **Step 2: Download and Read In Data**

This tutorials uses the movieless dataset. This dataset contains actual user ratings of movies.

In step 2, we will follow the steps below to get the datasets:
1. Go to https://grouplens.org/datasets/movielens/
2. Download the 100k dataset with the file name "ml-latest-small.zip"
3. Unzip "ml-latest-small.zip"
4. Copy the "ml-latest-small" folder to your project folder

Those who are using Google Colab for this analysis need to mount Google Drive to read the dataset. You can ignore the code below if you are not using Google Colab.



In [12]:
# Mount Google Drive
from google.colab import drive
drive.mount('/content/drive', force_remount=True)

Mounted at /content/drive


In [2]:
!pwd

/content


In [4]:
# Change directory
import os
os.chdir("drive/MyDrive/Recommendation_System")

# Print out the current directory
!pwd

/content/drive/MyDrive/Recommendation_System


There are multiple datasets in the 100k movielens folder. For this tutorial, we will use two ratings and movies.

Now let's read in the rating data.

In [7]:
import pandas as pd

In [8]:
# Read In data
ratings = pd.read_csv("ml-latest-small/ratings.csv")

# Take a look at the data
ratings.head()

userId  movieId  rating  timestamp
0       1        1     4.0  964982703
1       1        3     4.0  964981247
2       1        6     4.0  964982224
3       1       47     5.0  964983815
4       1       50     5.0  964982931

There are four columns in the ratings dataset, userID, movieID, rating, and timestamp.

The dataset has over 100k records, and there is no missing data.

In [9]:
# Get the dataset information
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100836 entries, 0 to 100835
Data columns (total 4 columns):
 #   Column     Non-Null Count   Dtype  
---  ------     --------------   -----  
 0   userId     100836 non-null  int64  
 1   movieId    100836 non-null  int64  
 2   rating     100836 non-null  float64
 3   timestamp  100836 non-null  int64  
dtypes: float64(1), int64(3)
memory usage: 3.1 MB


The 100k ratings are from 610 users on 9724 movies. The rating has ten unique values from 0.5 to 5.

In [13]:
# Number of users
print('The ratings dataset has', ratings['userId'].nunique(), 'unique users')

# Number of movies
print('The ratings dataset has', ratings['movieId'].nunique(), 'unique movies')

# List of unique ratings
print('The unique ratings are', sorted(ratings['rating'].unique()))

The ratings dataset has 610 unique users
The ratings dataset has 9724 unique movies
The unique ratings are [0.5, 1.0, 1.5, 2.0, 2.5, 3.0, 3.5, 4.0, 4.5, 5.0]


Next; Let's read in the movies data to get the movie names.

The movies dataset has movieID, title, and genres.

In [14]:
# Read in Data
movies = pd.read_csv('ml-latest-small/movies.csv')

# Take a look at the data
movies.head()

movieId                               title  \
0        1                    Toy Story (1995)   
1        2                      Jumanji (1995)   
2        3             Grumpier Old Men (1995)   
3        4            Waiting to Exhale (1995)   
4        5  Father of the Bride Part II (1995)   

                                        genres  
0  Adventure|Animation|Children|Comedy|Fantasy  
1                   Adventure|Children|Fantasy  
2                               Comedy|Romance  
3                         Comedy|Drama|Romance  
4                                       Comedy

In [15]:
# Get the dataset information
movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9742 entries, 0 to 9741
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   movieId  9742 non-null   int64 
 1   title    9742 non-null   object
 2   genres   9742 non-null   object
dtypes: int64(1), object(2)
memory usage: 228.5+ KB


Using 'movieID' as the matchmaking key, we appended movie information to the rating dataset and named it 'df'. So now we have the movie tile and movie rating in the same dataset!

In [16]:
# Merge ratings and movies datasets
df = pd.merge(ratings, movies, on='movieId', how='inner')

# Take a look at the data
df.head()

userId  movieId  rating   timestamp             title  \
0       1        1     4.0   964982703  Toy Story (1995)   
1       5        1     4.0   847434962  Toy Story (1995)   
2       7        1     4.5  1106635946  Toy Story (1995)   
3      15        1     2.5  1510577970  Toy Story (1995)   
4      17        1     4.5  1305696483  Toy Story (1995)   

                                        genres  
0  Adventure|Animation|Children|Comedy|Fantasy  
1  Adventure|Animation|Children|Comedy|Fantasy  
2  Adventure|Animation|Children|Comedy|Fantasy  
3  Adventure|Animation|Children|Comedy|Fantasy  
4  Adventure|Animation|Children|Comedy|Fantasy

## **Step 3: Exploratory Data Analysis (EDA)**